Identity basis (#282)
flatironinstitute/nemos: docs/how_to_guide/categorical_predictors.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

# Construct Design Matrices for Categorical Features

> ## Design Matrix with `patsy`

Reviewer: should be clear here that `patsy` is optional, like scikit learn, but that might be hlpeful

Author: I added a dropdown admonition

If you have categorized your trials and want to capture a change in the firing rate with the trial category, you can create your design matrix using [`patsy`](https://patsy.readthedocs.io/en/latest/).

Let's assume we have a simple dataset with four samples and two labels assigned to each sample. You can think of these labels as characteristics of the trial to which the samples belong, such as stimulus 1 vs stimulus 2 or context 1 vs context 2.

```{code-cell} ipython3
# Example data
import pandas as pd
from patsy import dmatrix
import numpy as np

# Example data 

> data = pd.DataFrame({

Reviewer: this example is very abstract. I think we'd better off using specific stimulus / context, and pointing out that the general principle applies

Author: I tried to make the stim and context more concrete, let me know if it works

Reviewer: I think this is good!

    'stimulus': ['s1', 's1', 's2', 's2'],  # Stimulus
    'context': ['c1', 'c2', 'c1', 'c2']   # Context
})

# Spike counts
counts = np.array([10, 5, 2, 0])
```

You can use a `patsy` formula to structure the design matrix:

```{code-cell} ipython3

# (a:b) indicates interaction between a and b
formula = "stimulus + context + stimulus:context"
```

To construct the design matrix:

```{code-cell} ipython3
# Generate design matrix
design_df = dmatrix(formula, data, return_type="dataframe")

design_df
```

The feature in the design matrix are 1-hot encoded categories.
Note that `patsy` adds an intercept and drops the reference `s1`. This is done by design to avoid perfect collinearity (the sum of the `s1` and the `s2` column would be equal to the intercept). 

:::{note} Understanding `patsy`'s Output
:class: dropdown

In the design matrix:
- `T.` indicates **treatment coding**, which is used to encode categorical variables. For example, `stimulus[T.s2]` represents the presence (`1`) or absence (`0`) of the category `s2` compared to the reference category `s1`.
- The reference category (`s1`) is dropped to avoid collinearity, as including all categories would result in redundancy.

> - Similarly, for `context`, columns `context[c1]` and `T.context[c2]` represent the presence of each category. 

Reviewer: ```suggestion
- Similarly, for `context`, columns `context[c1]` and `T.context[c2]` represent the presence of each category, and `context[T.c2]` is dropped to avoid collinearity. 
```

- The interaction term `stimulus[T.s2]:context[T.c2]` represents the combined effect of `stimulus = s2` and `context = c2`. Only one column is needed for the interaction, as the other combinations are implicitly represented by the reference categories (`s1` and `c1`).
:::
- 

> NeMoS GLMs, however, already specify an intercept term, making the first column of this dataframe redundant. Let's drop it.

Reviewer: ```suggestion
NeMoS GLMs, however, implicitly specify an intercept term, making the first column of this dataframe redundant. Let's drop it.
```

> 

Reviewer: this is rendered weirdly -- what's the `-` doing here? I also think this should be part of hte paragraph befor ethe admonition.

I think the admonition should be moved to right after the code box, the title needs to be corrected (it's currently not rendering as the title), and the admonition should probably have a link to the relevant part of the patsy docs: "See patsy docs for more details"

```{code-cell} ipython3
design_df.drop(columns=["Intercept"], inplace=True)

design_df
```

### Fit the GLM

> 

Reviewer: need some text here explaining what's happening

```{code-cell} ipython3
import nemos as nmo

# Fit the GLM model

> model = nmo.glm.GLM().fit(design_df, counts)

Reviewer: this is a bit confusing, because we don't talk about pandas dataframes anywhere else. I'm assuming it works because we just grab `design_df.values`, the array representation, but let's make that explicit.

Author: added an admonition about it

Author: let me know if it works

Reviewer: added some suggetions

```

## Categorical Predictors and Basis Functions

Assume you have an additional time series that you want to process with a basis:

```{code-cell} ipython3
# Example time series
speed = np.array([10., 3., 2., 20.])
```

> ### Composite Basis for Multiple Predictors

Reviewer: this header doesn't add anything

You can create a composite basis to combine predictors:

```{code-cell} ipython3
# Identity basis combined with a B-spline basis
bas = nmo.basis.IdentityEval() + nmo.basis.BSplineEval(5)

# Compute features

> X = bas.compute_features(design_df, speed)

Reviewer: should describe what this does: identity basis just passes through the design_df (again, I think we should add `.values`), while `BSplineEval` works with speed.

and probably link to the bit about additive bases

maybe show the array? or plot it?

Author: the `.values` is not needed, i think it's extra syntax that they should not care about. If the dataframe contains non-numerical values it will fail.

I'll add a description and plot the content

Reviewer: but it will fail if they try to do use a dataframe with non-numeric values anyway, right?

Reviewer: I actually don't think we should display it as a dataframe -- we should show what the output looks like: an appropriately shaped array.

```

### Interaction with Basis algebra

> To model the interaction between two variables, you can take advantage of the basis multiplication:

Reviewer: this is an additional interaction right? so maybe say something like 
```suggestion
To add the interaction between the `c2` context and speed to your design matrix, you can take advantage of basis multiplication:
```

```{code-cell} ipython3
# Interaction basis
bas = nmo.basis.IdentityEval() * nmo.basis.BSplineEval(5)

# Compute features for interaction
X2 = bas.compute_features(design_df["context[T.c2]"], speed)

> ```

Reviewer: and again, I think we should visualize this output

Author: I wrapped the output into a DataFrame to add back column names and have a more readable output

Reviewer: same point about not showing it as a dataframe. just show it as an array and explain what the columns mean in words

:color: info
:icon: info

In the design matrix:
- `T.` indicates **treatment coding**, which is used to encode categorical variables. For example, `stimulus[T.Sq]` represents the presence (`1`) or absence (`0`) of the category `Sq` compared to the reference category `Tri`.
- The reference category (`Tri`) is dropped to avoid collinearity, as including all categories would result in redundancy.
- Similarly, for `context`, columns `context[C]`, and `context[T.S]` is dropped to avoid collinearity.
- The interaction term `stimulus[T.Sq]:context[T.S]` represents the combined effect of `stimulus = Sq` and `context = S`. Only one column is needed for the interaction, as the other combinations are implicitly represented by the reference categories (`Tri` and `C`).

See [`patsy` docs](https://patsy.readthedocs.io/en/latest/formulas.html#the-formula-language) for more details. 
:::

### Fit the GLM

We are now ready to fit a GLM model using the 1-hot encoded categories as predictor. The model will learn a different 
firing rate for each condition. 

```{code-cell} ipython3
import nemos as nmo

# Fit the GLM model
model = nmo.glm.GLM().fit(design_df, counts)
```
:::{note}

> You can directly pass a [`pandas.DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) to the 
> `fit` method, as long as it can be internally converted to a floating-point array. The conversion is handled automatically, 
> so no additional steps are required on your part.

Reviewer: ```suggestion
You can directly pass a [`pandas.DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) to the 
`fit` method, as long as it can be internally converted to a floating-point array (that means, for example, that you can't have a column whose values are strings). The conversion is handled automatically, by calling `design_df.values`,
so no additional steps are required on your part.
```
